src/blog/speeding-up-v8-heap-snapshots.md

```diff
@@ -0,0 +1,161 @@
+
+[This hashing fix](https://chromium-review.googlesource.com/c/v8/v8/+/4428811) has landed in V8.
+
+## Second Optimization: Caching source positions
+
+After fixing the hashing, we re-profiled and found a further optimization opportunity that would take a significant part of the time. *An old friend*.
+
+For each allocation in the V8 heap, the heap snapshot tries to record the call stack responsible for the allocation. Therefore, for each stack frame, it needs to know the function name and its source location (filename, line number, column number). It was the fetching of this information that turned out to be super slow!
+
+What was happening? It was something similar to [what I fixed in the ETW stack walk](https://chromium-review.googlesource.com/c/v8/v8/+/3669698) and that I explained in [this post](https://blogs.igalia.com/dape/2022/12/21/native-call-stack-profiling-3-3-2022-work-in-v8/).
+
+To compute the source code lines of a function, V8 knows the linear position of the function in the script. But, to find the source line, it needs to traverse the whole script to identify where each newline occurs. This is expensive.
+
+When requesting line information, V8 already implements caching the source line positions per-script in method `Script::InitLineEnds()`. That information is stored in each script using a new heap object. Unfortunately, that was a problem that prevented the heap snapshot from using the cache. This is because, when the snapshot implementation traverses the heap, it cannot modify it, so the newly calculated line information cannot be cached.
+
+The solution? Before generating the heap snapshot, we now iterate over all the scripts in the V8 context to compute and cache the new source line positions information. As this is done out of the heap traversal done while capturing the snapshot, it is still possible to modify the heap to store the cached source code lines information.
+
+[This source position caching fix](https://chromium-review.googlesource.com/c/v8/v8/+/4538766) has also landed in V8.
+
+## Did we make it fast?
+
+After enabling both fixes, we re-profiled. Both of our fixes only affect snapshot generation time, so, as expected, snapshot serialization times were unaffected.
+
+When operating on a JS program containing...
+
+- Development JS, generation time is **50% faster** 👍
+- Production JS, generation time is **90% faster** 😮
+
+Why such a massive difference between production and development code? It is mostly because of the production source code optimizations. In production, code is optimized using bundling and minification, so there are fewer JS files and these files tend to be large, which means finding the newlines takes longer. And therefore, these larger files benefit most from the source position caching that only performs this work once.
```

Comment:
Some grammar and clarity fixes:

```suggestion
Why was there a massive difference between production and development code? The prodcution code is optimized using bundling and minification, so there are fewer JS files, and these files tend to be large. It takes longer to calculate source line positions for these large files, so they benefit the most when we can cache the source position and avoid repeated calculations.
```

Reply:
Applying after fixing the typo with prodctuion -> production